rust/ql/test/library-tests/dataflow/sources/test.rs

```diff
@@ -82,3 +82,130 @@ async fn test_reqwest() -> Result<(), reqwest::Error> {
 
     Ok(())
 }
+
+use std::io::Write;
+use http_body_util::BodyExt;
+
+async fn test_hyper_http(case: i64) -> Result<(), Box<dyn std::error::Error>> {
+    // using http + hyper libs to fetch a web page
+    let address = "example.com:80";
+    let url = "http://example.com/";
+
+    // create the connection
+    println!("connecting to {}...", address);
+    let stream = tokio::net::TcpStream::connect(address).await?;
+    let io = hyper_util::rt::TokioIo::new(stream);
+    let (mut sender, conn) = hyper::client::conn::http1::handshake(io).await?;
+
+    // drive the HTTP connection
+    tokio::task::spawn(async move {
+        conn.await.expect("connection failed");
+    });
+
+    // make the request
+    println!("sending request...");
+    if (case == 0) {
+        // simple flow case
+        let request = http::Request::builder().uri(url).body(String::from(""))?;
+        let mut response = sender.send_request(request).await?; // $ Alert[rust/summary/taint-sources]
+        sink(&response); // $ hasTaintFlow=request
+        sink(response); // $ hasTaintFlow=request
+        return Ok(())
```

Comment:
I've added this case to show some weird behaviour.  We get flow from the source, through `.await` (thanks to recent work) and `?` but it only reaches the sink if the response _isn't_ unpacked as in the original code (below).

I think we should treat solving this mystery as a follow-up task as it may well be a data flow issue entirely orthogonal to the modelling work (and besides, there's enough in this PR already).

Reply:
I've created an issue to track this.  I'm guessing this is something to do with borrowing, I'm very interested to here other people's theories.

Reply:
We probably do want to improve this at some point.  Real code is going to do this kind of thing quite a lot.